tools/ResxGen/ResxGen.psm1

```diff
@@ -0,0 +1,553 @@
+            }
+        }
+    }
+}
+
+function New-ResourceCode
+{
+    param
+    (
+        [Parameter(Mandatory)]
+        [ValidateNotNull()]
+        [Manifest] $manifest,
+
+        [Parameter(Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [string] $namespaceName,
+
+        [Parameter(Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [string] $className
+    )
+    $sb = [System.Text.StringBuilder]::new()
+    $null = $sb.AppendFormat($codePrologue, $namespaceName, $className, $manifest.FileName)
+    foreach ($eventEntry in $manifest.Events.Values)
+    {
+        $null = $sb.AppendFormat($codeEventEntryTemplate, $className, $eventEntry.EventId, $eventEntry.EventMessage.Id, $eventEntry.EventMessage.ParameterCount)
+    }
+    $null = $sb.Append($codeEpilogue)
+    $code = $sb.ToString().Replace('}}', '}')
+    return $code
+}
+
+<#
+.SYNOPSIS
+    Creates a resx file containing the messages from a manifest
+
+.PARAMETER messages
+    The EventMessage hash table containing the manifest messages
+#>
+function New-Resx
+{
+    param
+    (
+        [Manifest] $manifest
+    )
+    $messages = $manifest.Messages
+    $sb = [System.Text.StringBuilder]::new()
+
+    $null = $sb.AppendFormat($resxPrologue, $manifest.FileName)
+    foreach ($message in $messages.Values)
+    {
+        $null = $sb.AppendFormat($resxEntryTemplate, $message.Id, $message.Value)
+    }
+    $null = $sb.Append($resxEpilogue)
+    return $sb.ToString()
+}
+
+<#
+.SYNOPSIS
+    Generates a resx file and code file from an ETW manifest.
+
+.PARAMETER Manifest
+    The path to the ETW manifest file to read.
+
+.PARAMETER Name
+    The name to use for the C# class, the code file, and the resx file.
+    The default value is EventResource.
+
+.PARAMETER Namespace
+    The namespace to place the C# class.
+    The default is System.Management.Automation.Tracing.
+
+.PARAMETER ResxPath
+    The path to the directory to use to create the resx file.
+
+.PARAMETER CodePath
+    The path to the directory to use to create the C# code file.
+#>
+function ConvertTo-Resx
+{
+    [CmdletBinding()]
+    param
+    (
+        [Parameter(Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [string] $Manifest,
+
+        [string] $Name = 'EventResource',
+
+        [Parameter(Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [string] $Namespace = 'System.Management.Automation.Tracing',
+
+        [Parameter(Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [string] $ResxPath,
+
+        [Parameter(Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [string] $CodePath
+    )
+
+    [Manifest] $etwmanifest = [Manifest]::new($Manifest)
+
+    $resxFileName = Join-Path -Path $ResxPath -ChildPath "$($Name).resx"
+    Write-Verbose -Message "Creating $resxFileName" -Verbose
+
+    $resx = New-Resx -manifest $etwmanifest
+    $resx | Set-Content -Path $resxFileName -Encoding 'UTF8'
```

Comment:
We've moved to UTF8 with no BOM in this repo.

Reply:
What does that mean in practice?

Reply:
It's probably fine to just use ASCII when generating this file.

Reply:
Are you referring to the code or resx file?  ASCII would preclude localization of the resx file; wouldn't it?

Reply:
That's just a shortcut assuming that we don't have characters requiring UTF8 encoding, which we don't.

This would only preclude localization if we generated the resx file from a localized manifest. As we're not even localizing error messages, I think we're a long way off from localizing the event log, and when that happens, it might not depend on generating new resx files - so I wouldn't worry about that for now.

Reply:
Fair enough, I'll switch to ascii.

Reply:
Fixed.